assembly/src/assembler/procedure_cache.rs

```diff
@@ -0,0 +1,62 @@
+use super::{btree_map::Entry, BTreeMap, Procedure, ProcedureId, RpoDigest};
+
+#[cfg(test)]
+use super::btree_map::Values;
+
+/// The [ProcedureCache] is responsible for caching [Procedure]s. It allows [Procedure]s to be
+/// fetched using both [ProcedureId] and [RpoDigest].
+#[derive(Debug, Default)]
+pub struct ProcedureCache {
+    proc_id_map: BTreeMap<ProcedureId, RpoDigest>,
+    mast_map: BTreeMap<RpoDigest, Procedure>,
+}
+
+impl ProcedureCache {
+    // ACCESSORS
+    // --------------------------------------------------------------------------------------------
+    /// Returns a [Procedure] reference corresponding to the [ProcedureId].
+    pub fn get_by_id(&self, id: &ProcedureId) -> Option<&Procedure> {
+        self.proc_id_map.get(id).and_then(|mast_hash| self.mast_map.get(mast_hash))
+    }
+
+    /// Returns a [Procedure] referene corresponding to the MAST root ([RpoDigest]).
+    pub fn get_by_hash(&self, root: &RpoDigest) -> Option<&Procedure> {
+        self.mast_map.get(root)
+    }
+
+    /// Returns true if the [ProcedureCache] contains a [Procedure] for the specified [ProcedureId].
+    pub fn contains_id(&self, id: &ProcedureId) -> bool {
+        self.proc_id_map.contains_key(id)
+    }
+
+    /// Returns true if the [ProcedureCache] contains a [Procedure] for the specified root
+    /// ([RpoDigest]).
+    pub fn contains_hash(&self, root: &RpoDigest) -> bool {
+        self.mast_map.contains_key(root)
+    }
+
+    /// Returns an iterator over the [Procedure]s in the [ProcedureCache].
+    #[cfg(test)]
+    pub fn values(&self) -> Values<RpoDigest, Procedure> {
```

Comment:
Should we change return type here to `impl Iterator<Item = &Procedure>`? This way, we don't need to import `Values` from `btree_map`.